zebra-state/src/service/non_finalized_state/chain.rs

```diff
@@ -355,12 +355,42 @@ impl UpdateWith<Option<transaction::JoinSplitData<Groth16Proof>>> for Chain {
         joinsplit_data: &Option<transaction::JoinSplitData<Groth16Proof>>,
     ) -> Result<(), ValidateContextError> {
         if let Some(joinsplit_data) = joinsplit_data {
-            for sprout::JoinSplit { nullifiers, .. } in joinsplit_data.joinsplits() {
-                let span = debug_span!("revert_chain_state_with", ?nullifiers);
+            for nullifier in joinsplit_data.nullifiers() {
+                let span = debug_span!("update_chain_state_with", ?nullifier);
                 let _entered = span.enter();
-                trace!("Adding sprout nullifiers.");
-                self.sprout_nullifiers.insert(nullifiers[0]);
-                self.sprout_nullifiers.insert(nullifiers[1]);
+                trace!("adding sprout nullifier");
+
+                // Reject double-spends of nullifers:
+                // - both within the same JoinSplit,
+                // - from different JoinSplits in this transaction's JoinSplitData, or
+                // - one from this joinsplit, and another from:
+                //   - a previous transaction in this Block, or
+                //   - a previous block in this non-finalized Chain.
+                //
+                // (Duplicate finalized nullifiers are rejected during contextual validation.)
+                //
+                // TODO:
+                // - test that the chain's nullifiers are not modified on error
+                //
+                // "A nullifier MUST NOT repeat either within a transaction,
+                // or across transactions in a valid blockchain.
+                // Sprout and Sapling and Orchard nullifiers are considered disjoint,
+                // even if they have the same bit pattern."
+                //
+                // https://zips.z.cash/protocol/protocol.pdf#nullifierset
```

Comment:
Do you think I should just delete this consensus rule, and go with the shorter one?

Reply:
I ended up putting the long consensus rule at the top of the `nullifier` module, and then quoting the short or specific consensus rules for each function. See ef1d2fb4.